### PR TITLE
fix(docs-infra): improve cancel icon in top-menu search

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
webkit browsers show a cancel X icon in the input search on focus/hover
which allows to clear the content of the input, such icon has small
glitches/imperfections so use a mat-icon instead

(this also adds such icon to non-webkit browsers)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

This related to the cancel icon in the top-menu search input (present in webkit browsers):
Screenshot at 2021-09-26 13-08-05

I often see quirky behaviors, like it being partially "cut" (this may need changing the browser's font-size to reproduce): 
Screenshot at 2021-09-26 13-07-00

Or very often, leaving "marks" after disappearing:
Screenshot at 2021-09-26 13-05-46

## What is the new behavior?

The icon has been hidden and it's been replaced with the cancel mat-icon:
Screenshot at 2021-09-26 13-07-50
(to me it even looks more slick/modern :slightly_smiling_face: )

This gives us more control on the icon allowing us to fix the aforementioned quirks:

Screenshot at 2021-09-26 13-07-23

Screenshot at 2021-09-26 13-12-28

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
 - I appreciated that the quirks mentioned above may have happened only for me.... if that's the case please let me know :sweat: 
 (My browser is: Chormium Version 93.0.4577.82 (Official Build) snap (64-bit))
 - this change also adds the icon to non webkit browsers (like firefox), not sure if this is ok or not, please let me know :slightly_smiling_face: 
 - I could only test this changes in Chromium and Firefox, hopefully they look ok on other browsers too :sweat: 
 - this is a pretty minor thing so it may not even be worth adding a new mat-icon and css/logic for it, but I just wanted to open this PR as the marks left behind have been a bit of an eye sore for me for a while :sweat_smile:

### CODE REVIEW

File: aio/src/app/search/search-box/search-box.component.ts [hunk 1]
```diff
@@ -22,7 +22,11 @@ import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
     (input)="doSearch()"
     (keyup)="doSearch()"
     (focus)="doFocus()"
-    (click)="doSearch()">`
+    (click)="doSearch()">
+    <mat-icon
```
Comment (dario-piotrowicz) [inline aio/src/app/search/search-box/search-box.component.ts:26]: to keep things simple I just added this `mat-icon` and styled it using the tag name, I could add a class like `cancel-icon` and use that instead, not too sure what's the preference here :slightly_smiling_face:

File: goldens/size-tracking/aio-payloads.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2017": 4841,
-        "main-es2017": 462310,
+        "main-es2017": 462962,
```
Comment (dario-piotrowicz) [inline goldens/size-tracking/aio-payloads.json:6]: @gkalpak :sweat: :cry:

File: aio/src/styles/1-layouts/top-menu/_top-menu.scss [hunk 3]
```diff
@@ -189,6 +191,25 @@ mat-toolbar.app-toolbar {
       @media (max-width: 515px) {
         width: 150px;
       }
+
+      &::-webkit-search-cancel-button {
+        display: none;
+      }
+
+      &:not(:focus-within,:hover) + mat-icon:not(:hover,:focus) {
```
Comment (dario-piotrowicz) [inline aio/src/styles/1-layouts/top-menu/_top-menu.scss:199]: `:focus-within` doesn't seem necessary `:focus` should be enough

also:
Screenshot at 2021-09-26 16-16-55

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx for working on this, @dario-piotrowicz!

As you said, this is really a minor issue, so I am not concerned much. And yes, the custom icon does look slightly more modern/fitting. The payload size increase is not great for sure, but it is quite small (percentage-wise), so not too concerning either.

FWIW, I have never seen the left-over marks you mentioned 🤷‍♂️

All in all, I wouldn't be opposed to getting this is, if we can address a couple of downsides I noticed (without imposing a significant payload size increase):
- On IE, two clear buttons are shown. (According to [this article][1], you could get rid of the built-in "clear" icon with `::-ms-clear`.)
- Unlike what happens with the built-in icon, the custom icon can overlay text (if you enter a long enough search query), which slightly hurts readability.
- Unlike what happens with the built-in icon, the custom icon does not return focus to the search input when clicked, which slightly hurts usability.

[1]: https://blog.maximerouiller.com/post/remove-the-x-from-internet-explorer-and-chrome-input-type-search

File: goldens/size-tracking/aio-payloads.json (hunk 2 above)
Comment (gkalpak) [reply goldens/size-tracking/aio-payloads.json:6]: 😟 :smiley:

**Comment by dario-piotrowicz** [comment]
> Thx for working on this, @dario-piotrowicz!
> 
> As you said, this is really a minor issue, so I am not concerned much. And yes, the custom icon does look slightly more modern/fitting. The payload size increase is not great for sure, but it is quite small (percentage-wise), so not too concerning either.
> 
> FWIW, I have never seen the left-over marks you mentioned man_shrugging
> 
> All in all, I wouldn't be opposed to getting this is, if we can address a couple of downsides I noticed (without imposing a significant payload size increase):
> 
> * On IE, two clear buttons are shown. (According to [this article](https://blog.maximerouiller.com/post/remove-the-x-from-internet-explorer-and-chrome-input-type-search), you could get rid of the built-in "clear" icon with `::-ms-clear`.)
> * Unlike what happens with the built-in icon, the custom icon can overlay text (if you enter a long enough search query), which slightly hurts readability.
> * Unlike what happens with the built-in icon, the custom icon does not return focus to the search input when clicked, which slightly hurts usability.

I see.... yeah I would not want to fix something that is not broken... I have no idea to how many people this happens... maybe be my browsers/os or whatever... :sweat_smile: but if you're not opposed to it I guess I will try to address the issues you presented and we can take it from there :slightly_smiling_face:

**Comment by dario-piotrowicz** [comment]
@gkalpak I amended what you mentioned, now clicking on the icon focuses in the input element :slightly_smiling_face: 

Also I added right padding so that the text does not go under the icon (note: if you remove focus/hover from the input this padding remains, I think this is ok as this is the same behaviour as the current webkit implementation)

About IE, I put that fix too, but I could not test that, I was wondering if you could have a look at it for me :sweat: (if you can't I'll have to mess around with VMs and stuff like that to get IE on my linux machine :sweat:)

**Comment by dario-piotrowicz** [comment]
My latest implementation (at least for me) seems to work exactly as the webkit implementation now :slightly_smiling_face: , the only thing different being that the webkit icon doesn't have `cursor: pointer` but the new icon does, I like it better this was, feels more polished, but if you disagree I can remove it :+1:

File: aio/src/styles/1-layouts/top-menu/_top-menu.scss [hunk 4]
```diff
@@ -189,6 +191,29 @@ mat-toolbar.app-toolbar {
...
         width: 150px;
       }
+
+      &::-webkit-search-cancel-button {
+        display: none;
+      }
+
+      &::-ms-clear {
+        display: none;
+      }
```
Comment (dario-piotrowicz) [inline aio/src/styles/1-layouts/top-menu/_top-menu.scss:201]: I tried doing:

```scss
      &::-webkit-search-cancel-button, &::-ms-clear {
        display: none;
      }
```

but for some reason that would not work and I would see two icons on Chrome :sweat:

File: aio/src/styles/1-layouts/top-menu/_top-menu.scss [hunk 5]
```diff
@@ -159,11 +160,12 @@ mat-toolbar.app-toolbar {
     min-width: 150px;
     height: 100%;
     margin-right: 16px;
+    position: relative;
 
     input {
       border: none;
       border-radius: 100px;
-      padding: 5px 16px;
+      padding: 5px 3rem 5px 16px;
```
Comment (dario-piotrowicz) [inline aio/src/styles/1-layouts/top-menu/_top-menu.scss:168]: @gkalpak is is ugly to mix `rem` and `px` like this? :thinking:

Comment (gkalpak) [reply aio/src/styles/1-layouts/top-menu/_top-menu.scss:168]: It's fine :smiley:

**Review by gkalpak — CHANGES_REQUESTED** [review]
LGTM (as soon as CI is green - currently it needs updating the [`aio-local` payload size limit for `styles`][1]).

[1]: https://github.com/angular/angular/blob/12d9e6e1a67c102dc515b02943aebe8562721098/goldens/size-tracking/aio-payloads.json#L20

**Comment by dario-piotrowicz** [comment]
> LGTM (as soon as CI is green - currently it needs updating the [`aio-local` payload size limit for `styles`](https://github.com/angular/angular/blob/12d9e6e1a67c102dc515b02943aebe8562721098/goldens/size-tracking/aio-payloads.json#L20)).

Yeah I noticed that but in the PR you can see I am indeed updating the value... could it be that CI didn't picked up on that? :thinking: 

anyways I'll just rebase later and see if that fixes it :slightly_smiling_face:

**Comment by gkalpak** [comment]
> Yeah I noticed that but in the PR you can see I am indeed updating the value... could it be that CI didn't picked up on that? 🤔

You are updating the `aio` limits, not the `aio-local` limits :wink:
(See the link in my previous comment.)

**Comment by dario-piotrowicz** [comment]
> > Yeah I noticed that but in the PR you can see I am indeed updating the value... could it be that CI didn't picked up on that? thinking
> 
> You are updating the `aio` limits, not the `aio-local` limits wink (See the link in my previous comment.)

oops sorry :sweat_smile:

By the way, what's the difference between `aio` and `aio-local`, we there are two? :thinking: 

Also just wanted to ask, so you checked and can confirm that it works fine on IE? :slightly_smiling_face:

**Review by gkalpak — APPROVED** [review]

**Comment by gkalpak** [comment]
> By the way, what's the difference between `aio` and `aio-local`, we there are two? 🤔

- The `aio` payload size entry corresponds to the regular AIO build (using the Angular package versions specified in `aio/package.json`). On CI, this is checked in the [`test_aio` job][1].
- The `aio-local` payload size entry corresponds to a special AIO build mode, where we build the Angular packages from sources and use those in AIO. On CI, this is checked in the [`test_aio_local` CI job][2].

> Also just wanted to ask, so you checked and can confirm that it works fine on IE? 🙂

Sorry, forgot to mention that it works OK on IE (as OK as things work on IE anyway).

[1]: https://github.com/angular/angular/blob/263feba5c2d56e8433068718d1fdcbc3b2ae144c/.circleci/config.yml#L399
[2]: https://github.com/angular/angular/blob/263feba5c2d56e8433068718d1fdcbc3b2ae144c/.circleci/config.yml#L432

**Review by petebacondarwin — APPROVED** [review]
~500 bytes seems reasonable.


Reviewed-for: size-tracking

**Review by IgorMinar — APPROVED** [review]
LGTM, thank you!

Reviewed-for: size-tracking

**Pull request closed without merging**